src/App.tsx

```diff
@@ -1,24 +1,196 @@
+  }, [todos]);
+
+  const visibleTodos = getFilteredTodos(todos, filterType);
+
+  const isSomeTodosCompleted: boolean = todos.some(item => item.completed);
+
+  const isAllTodosCompleted: boolean = todos.every(item => item.completed);
+
+  const handleTogglingTodo = useCallback(async (
+    todoId: number,
+    todoStatus: boolean,
+  ) => {
+    setLoadingTodoIds(prevIds => ([...prevIds, todoId]));
+    try {
+      await updateTodoStatus(todoId, todoStatus);
+      setErrorMessage(ErrorMessages.NoError);
+
+      setTodos((prevTodos: Todo[]): Todo[] => {
+        return prevTodos.map((item: Todo) => {
+          if (item.id === todoId) {
+            item.completed = todoStatus; // eslint-disable-line no-param-reassign
+          }
+
+          return item;
+        });
+      });
+    } catch (error) {
+      setErrorMessage(ErrorMessages.OnUpdate);
+    }
+
+    setLoadingTodoIds(prevIds => prevIds.filter(prevId => prevId !== todoId));
+  }, []);
+
+  const updateTitle = useCallback(async (
+    todoId: number,
+    title: string,
+  ) => {
+    setLoadingTodoIds(prevIds => ([...prevIds, todoId]));
+    try {
+      await updateTodoTitle(todoId, title);
+      setErrorMessage(ErrorMessages.NoError);
+
+      setTodos((prevTodos: Todo[]): Todo[] => {
+        return prevTodos.map((item: Todo) => {
+          if (item.id === todoId) {
+            item.title = title; // eslint-disable-line no-param-reassign
+          }
+
+          return item;
+        });
+      });
+    } catch (error) {
+      setErrorMessage(ErrorMessages.OnUpdate);
+    }
+
+    setLoadingTodoIds(prevIds => prevIds.filter(prevId => prevId !== todoId));
+  }, []);
+
+  const toggleAll = useCallback(() => {
+    let newTodoStatus = true;
+    let todosToUpdate = todos;
+
+    if (isAllTodosCompleted) {
+      newTodoStatus = false;
+    } else if (isSomeTodosCompleted) {
+      newTodoStatus = true;
+      todosToUpdate = getFilteredTodos(todos, FilterOptions.Active);
+    }
+
+    todosToUpdate.forEach(todo => handleTogglingTodo(todo.id, newTodoStatus));
+  }, [todos]);
+
+  const loadTodos = async () => {
+    try {
+      setErrorMessage(ErrorMessages.NoError);
+      const todosFromServer = await getTodos(USER_ID);
+
+      setTodos(todosFromServer);
+    } catch (error) {
+      setErrorMessage(ErrorMessages.OnLoad);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
```

Comment:
lets add Loader here, while we waiting for todos loaded

Reply:
I don't see it in the task.
Maybe, you mean this optional smooth animation from the previous task https://prnt.sc/ZM7lkm9Vzkjr ?
(https://github.com/mate-academy/react_todo-app-add-and-delete#if-you-want-to-implement-smooth-animations)
Or did I just get you wrong?

Reply:
hmm, no I mean Loader it's really strange why you haven't default loader in this task 
so ok you can skip this comment or just create Loader component


Loder
```
import React from 'react';
 import './Loader.scss';
 export const Loader: React.FC = React.memo(() => (
   <div className="Loader">
     <div className="Loader__content" />
   </div>
 ));
```

Styles
```
.Loader {
   display: flex;
   width: 100%;
   justify-content: center;
   align-items: center;

   &__content {
     border-radius: 50%;
     width: 2em;
     height: 2em;
     margin: 1em auto;
     border: 0.3em solid #ddd;
     border-left-color: #000;
     animation: load8 1.2s infinite linear;
   }
 }

 @keyframes load8 {
   0% {
     transform: rotate(0deg);
   }
   100% {
     transform: rotate(360deg);
   }
 }
```